hooks/nvidia-bootstrap/image/run.sh

```diff
@@ -12,50 +12,220 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-#!/bin/sh
+#!/bin/bash
+set -euo pipefail
 
-# Simple early detection of nvidia card
-grep 10de102d /proc/bus/pci/devices || exit 0
+#################################################
+# Settings
 
-# p2.xlarge
-# 00f0	10de102d	4b	        84000000	      100000000c	               0	        8200000c	               0	               0	               0	         1000000	       400000000	               0	         2000000	               0	               0	               0	nvidia
+# A place on the host machine to cache these huge 1.6GB+ downloads in between reboots.
+ROOTFS_DIR=/rootfs
+CACHE_DIR_HOST=/nvidia-bootstrap-cache
+CACHE_DIR_CONTAINER="${ROOTFS_DIR}${CACHE_DIR_HOST}"
 
+# AWS Instance Types to Nvidia Card Mapping (cut and pasted from AWS docs)
+# Load the correct driver for the correct instance type
+#   Instances  Product Type  Product Series  Product
+#   G2         GRID          GRID Series     GRID K520   <-- I think they meant G3
```

Comment:
This is correct, but I think G2 is a deprecated instance type. It no longer appears on the pricing page, though we have our own G2 instances running.

According to https://aws.amazon.com/ec2/instance-types/ and https://aws.amazon.com/ec2/instance-types/g3/, G3 instances are based on Tesla M60s.

Reply:
@115100  Thanks for the clarifications.

So, based on what you said:

1) We don't have to worry about G2 instances because one cannot spin them up because of deprecation (am I wrong?)

2) The driver package for the G3 instance is suboptimal but working (I tested it).  I'm looking at the Nvidia website right now and Tesla M60 hardware resolves to the following driver, which matches that of P2 and P3.

http://us.download.nvidia.com/tesla/390.46/NVIDIA-Linux-x86_64-390.46.run

I can make the doc change and driver swap later after more feedback, and if we think this PR has any chance of merging.

Reply:
G2 instances can still be launched but Nvidia's own drivers don't support kernels >=4.9 - I had to write my own patch to get the driver to work. Personally, I think it's safe to ignore G2.

On G3, I don't have any instances up to test them but the rest of the PR looks good to me.

Reply:
> Nvidia's own drivers don't support kernels >=4.9

That's not true. You didn't pick the right driver package then.

Reply:
Ah, I see a new set of drivers. It was a while back and it did take a while to support the newer kernel though.